cranelift/module/src/module.rs

```diff
@@ -558,13 +584,19 @@ pub trait Module {
     /// Returns the size of the function's code.
     fn define_function_bytes(
         &mut self,
-        func: FuncId,
+        func_id: FuncId,
+        func: &ir::Function,
         bytes: &[u8],
         relocs: &[MachReloc],
     ) -> ModuleResult<ModuleCompiledFunction>;
 
     /// Define a data object, producing the data contents from the given `DataContext`.
-    fn define_data(&mut self, data: DataId, data_ctx: &DataContext) -> ModuleResult<()>;
+    fn define_data(
+        &mut self,
+        data: DataId,
+        func: &ir::Function,
```

Comment:
???? Why do you need a Function to define a data object?

Reply:
It was to process relocations (to extract the , but now I can see that `DataDescription` abuses the `MachReloc` and it should be its own type that doesn't use `ir::ExternalName` at all, now that the latter uses references into an external table for lookup. I'll try doing this quickly, but if it takes too long I'd prefer that somebody who understands this code better take a look.

Reply:
That wasn't too hard, eventually; there's now a `ModuleReloc` that is a `MachReloc` but using `ModuleExtName` (which embeds user-defined name) instead of `ir::ExternalName` for names.